tikv/raftstore/snap_applier.go

```diff
@@ -0,0 +1,251 @@
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		it.writeCFIterator, err = rocksdb.NewSstFileIterator(it.writeCFFile)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		it.writeCFIterator.SeekToFirst()
+		if !it.writeCFIterator.Valid() {
+			return nil, it.writeCFIterator.Err()
+		}
+		it.curWriteKey, it.curWriteCommitTS, err = decodeRocksDBSSTKey(it.writeCFIterator.Key().UserKey)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+	}
+	return it, nil
+}
+
+func (ai *snapApplier) next() (*applySnapItem, error) {
+	if ai.curLockKey != nil && ai.curWriteKey != nil {
+		if bytes.Compare(ai.curLockKey, ai.curWriteKey) <= 0 {
+			return ai.nextLock()
+		} else {
+			return ai.nextWrite()
+		}
+	} else if ai.curLockKey != nil {
+		return ai.nextLock()
+	} else if ai.curWriteKey != nil {
+		return ai.nextWrite()
+	} else {
+		return nil, nil
+	}
+}
+
+func (ai *snapApplier) nextLock() (*applySnapItem, error) {
+	item := new(applySnapItem)
+	item.key = ai.curLockKey
+	item.applySnapType = applySnapTypeLock
+	item.useMeta = mvcc.LockUserMetaNone
+	lv, err := decodeLockCFValue(ai.curLockValue)
+	if err != nil {
+		return nil, err
+	}
+	val, err := ai.loadFullValue(item.key, lv.startTS, lv.shortVal, lv.lockType)
+	if err != nil {
+		return nil, err
+	}
+	mvccLock := new(mvcc.MvccLock)
+	mvccLock.Op = lv.lockType
+	mvccLock.StartTS = lv.startTS
+	mvccLock.TTL = uint32(lv.ttl)
+	mvccLock.PrimaryLen = uint16(len(lv.primary))
+	mvccLock.Primary = lv.primary
+	mvccLock.Value = val
+	err = ai.loadOldValue(mvccLock)
+	if err != nil {
+		return nil, err
+	}
+	item.val = mvccLock.MarshalBinary()
+	if len(ai.lockCFData) > 0 {
+		ai.curLockKey, ai.curLockValue, ai.lockCFData, err = readEntryFromPlainFile(ai.lockCFData)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		ai.curLockKey = nil
+	}
+	return item, err
+}
+
+func (ai *snapApplier) loadFullValue(key []byte, startTS uint64, shortVal []byte, op byte) ([]byte, error) {
+	if shortVal == nil && op == byte(kvrpcpb.Op_Put) {
+		if !ai.defaultCFIterator.Valid() {
+			return nil, errors.WithStack(errInvalidSnapshot)
+		}
+		defKey, defStartTS, err := decodeRocksDBSSTKey(ai.defaultCFIterator.Key().UserKey)
+		if err != nil {
+			return nil, err
+		}
+
+		if !bytes.Equal(key, defKey) {
+			return nil, errors.WithStack(errInvalidSnapshot)
+		}
+		if defStartTS != startTS {
+			return nil, errors.WithStack(errInvalidSnapshot)
+		}
+		val := y.SafeCopy(nil, ai.defaultCFIterator.Value())
+		ai.defaultCFIterator.Next()
+		return val, nil
+	}
+	return shortVal, nil
+}
+
+func (ai *snapApplier) loadOldValue(lock *mvcc.MvccLock) error {
+	if lock.Op == byte(kvrpcpb.Op_Lock) {
+		return nil
+	}
+	for bytes.Equal(ai.curLockKey, ai.curWriteKey) {
+		var err error
+		writeVal := decodeWriteCFValue(y.SafeCopy(nil, ai.writeCFIterator.Value()))
+		if writeVal.writeType == byte(kvrpcpb.Op_Rollback) {
+			// Skip rollback entry, find the next write key.
```

Comment:
How about if there is a prewrite in default cf and a rollback in write cf?

Reply:
The key in the default CF will be deleted when the rollback is written.

Reply:
Also need to skip the related value in default cf.

Reply:
I mean the default CF key is already deleted when the rollback is written in TiKV.

Reply:
OK